src/probability/borel_cantelli.lean

```diff
@@ -0,0 +1,133 @@
+/-
+Copyright (c) 2022 Kexing Ying. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Kexing Ying
+-/
+import probability.martingale.borel_cantelli
+import probability.independence
+
+/-!
+
+# The second Borel-Cantelli lemma
+
+This file contains the second Borel-Cantelli lemma which states that, given a sequence of
+independent sets `(sₙ)` in a probability space, if `∑ n, μ sₙ = ∞`, then the limsup of `sₙ` has
+measure 1. We employ a proof using Lévy's generalized Borel-Cantelli by choosing an appropriate
+filtration.
+
+## Main result
+
+- `probability_theory.measure_limsup_eq_one`: the second Borel-Cantelli lemma.
+
+-/
+
+open_locale measure_theory probability_theory ennreal big_operators topological_space
+
+open measure_theory probability_theory measurable_space
+
+namespace probability_theory
+
+variables {Ω : Type*} {m0 : measurable_space Ω} {μ : measure Ω} {s : ℕ → set Ω}
+
+section pi_system
+
+lemma Indep_set.indep_generate_from_le [is_probability_measure μ]
+  (hsm : ∀ n, measurable_set (s n)) (hs : Indep_set s μ) (n : ℕ) :
+  indep (generate_from {s (n + 1)}) (generate_from {t | ∃ k ≤ n, s k = t}) μ :=
+begin
+  have := hs.indep_generate_from_of_disjoint _ hsm {n + 1} {k | k ≤ n} (λ x ⟨hx₁, hx₂⟩,
+    not_lt.2 (set.mem_set_of.1 hx₂) ((set.mem_singleton_iff.1 hx₁).symm ▸ nat.lt_succ_self n)),
+  convert this,
+  simp only [set.mem_singleton_iff, exists_prop, exists_eq_left, set.set_of_eq_eq_singleton'],
+end
+
+end pi_system
+
+section borel_cantelli
+
+/-- Given a sequence of measurable sets `(sₙ)`, `filtration_of_set` is the smallest filtration
+such that `sₙ` is measurable with respect to the `n`-the sub-σ-algebra in `filtration_of_set`. -/
+def filtration_of_set {s : ℕ → set Ω} (hsm : ∀ n, measurable_set (s n)) : filtration ℕ m0 :=
```

Comment:
If we see the sets as functions (I'm not saying that we should do that in the code), this is the same thing as `filtration.natural`, isn't it?
Could you define this next to `filtration.natural`, and in a similar generality? This definition will be useful more generally than in this file.

Reply:
Done!